tractor/_broadcast.py

```diff
@@ -0,0 +1,275 @@
+                raise trio.ClosedResourceError
+
+            event = trio.Event()
+            state.recv_ready = key, event
+
+            try:
+                value = await self._recv()
+
+                # items with lower indices are "newer"
+                state.queue.appendleft(value)
+
+                # broadcast new value to all subscribers by increasing
+                # all sequence numbers that will point in the queue to
+                # their latest available value.
+
+                # don't decrement the sequence for this task since we
+                # already retreived the last value
+
+                # XXX: which of these impls is fastest?
+
+                # subs = state.subs.copy()
+                # subs.pop(key)
+
+                for sub_key in filter(
+                    # lambda k: k != key, state.subs,
+                    partial(ne, key), state.subs,
+                ):
+                    state.subs[sub_key] += 1
+
+                return value
+
+            finally:
+                # reset receiver waiter task event for next blocking condition
+                event.set()
+                state.recv_ready = None
+
+        # This task is all caught up and ready to receive the latest
+        # value, so queue sched it on the internal event.
+        else:
+            _, ev = state.recv_ready
+            await ev.wait()
+            seq = state.subs[key]
+            state.subs[key] -= 1
+            return state.queue[seq]
+
+    @asynccontextmanager
+    async def subscribe(
+        self,
+    ) -> AsyncIterator[BroadcastReceiver]:
+        '''Subscribe for values from this broadcast receiver.
+
+        Returns a new ``BroadCastReceiver`` which is registered for and
+        pulls data from a clone of the original ``trio.abc.ReceiveChannel``
+        provided at creation.
+
+        '''
+        # use a uuid4 for a tee-instance token
+        key = str(uuid4())
+        state = self._state
+        br = BroadcastReceiver(
+            key=key,
+            rx_chan=self._rx,
+            state=state,
+            receive_afunc=self._recv,
+        )
+        # assert clone in state.subs
+        assert key in state.subs
+
+        try:
+            yield br
+        finally:
+            await br.aclose()
+
+    # TODO:
+    # - should there be some ._closed flag that causes
+    #   consumers to die **before** they read all queued values?
+    # - if subs only open and close clones then the underlying
+    #   will never be killed until the last instance closes?
+    #   This is correct right?
+    async def aclose(
+        self,
+    ) -> None:
+
+        if self._closed:
+            return
+
+        # XXX: leaving it like this consumers can still get values
+        # up to the last received that still reside in the queue.
+        # Is this what we want?
+        self._state.subs.pop(self.key)
+        # if not self._state.subs:
+        #     await self._rx.aclose()
```

Comment:
This shouldn't get touched right?

It's up to the caller to determine when the underlying should be closed?

Reply:
Yeah sticking with underlying lifetime mgmt being up to use code.